Initial event-native documentation (#935)
gravitee-io/gravitee-docs: pages/apim/3.x/v4/v4-introduction.adoc

:page-permalink: apim/3.x/v4_introduction.html
:page-folder: apim/event-native-apim
:page-layout: apim3x

[label label-version]#New in version 3.20.0#
[label label-version]#BETA release#

Introduced in release 3.19.0 of the Gravitee API Management platform, there is a beta version of a new policy execution engine that enables an improved execution flow for synchronous APIs and supports event-driven policy execution for asynchronous APIs. It is based on a modern and fully reactive architecture designed to address a number of challenges Gravitee users have been facing with the existing policy execution engine, available with the Gravitee's V2 API definition specification used in versions prior to APIM 3.19.x.

The new policy execution engine builds on the new V4 BETA API definition foundation of Gravitee's event-native platform, which adds features such as native support for Pub/Sub (Publish-Subscribe) design and enabling policies at message level. The new V4 BETA event-native API definition makes it possible for the Gateway to “know” what is necessary in order for it to handle asynchronous API use cases.

The new engine provides the following capabilities:

> * The ability to execute policies in the exact order in which they have been placed in Policy Studio. This addresses some issues experienced by users related to the order in which policies are executed by the V3 engine where policies interacting with the Head part of the request are always executed first, even when placed in a different order in the Policy Studio during the design phase. With the new execution engine, it is possible to apply logic on a head policy based on the payload of the request - for example, to apply dynamic routing based on the request payload.

Reviewer: ```suggestion
* The ability to execute policies in the exact order in which they have been placed in Design Studio. This addresses some issues experienced by users related to the order in which policies are executed by the V3 engine where policies interacting with the Head part of the request are always executed first, even when placed in a different order in the Design Studio during the design phase. With the new execution engine, it is possible to apply logic on a head policy based on the payload of the request - for example, to apply dynamic routing based on the request payload.
```

Reviewer: @mouligno we mean Design Studio here, right?
@AndrewMcFarlandCampbellGravitee we could add the xref link to the relevant doc pages e.g. https://docs.gravitee.io/apim/3.x/apim_publisherguide_design_studio_create.html#flow-policies and https://docs.gravitee.io/apim/3.x/apim_publisherguide_plan_policies.html

* Proper isolation between platform-level policies and API-level policies during policy execution. This ensures that platform-level policies are always executed prior to any API-level policies during the request stage, and after any API-level policies during the response stage.
* Removal of the need to define a scope for policies (`onRequest`, `onRequestContent`, `onResponse`, `onResponseContent`).
* The introduction of new API types to better differentiate REST APIs from Async APIs (Websocket, SSE, Webhook), as a foundation to fully support the execution of Async APIs.

> NOTE: This BETA release does not provide GUI support for these capabilities. You can check out the new engine behaviour with the  link:{{ 'apim/3.x/event_native_apim_example_use_cases.html' | relative_url }}[example use cases].

Reviewer: ```suggestion
NOTE: This BETA release does not provide GUI support for these capabilities. You can check out the new engine behaviour with the  link:{{ 'apim/3.x/event_native_apim_example_use_cases.html' | relative_url }}[V4 BETA event-native API Management example use cases].
```

> WARNING: The new policy execution engine mode has been released as a BETA version: it should not be used in production. The GA, due in 2023, release will come with additional features - link:https://www.gravitee.io/contact-us[contact us] if you want to know more about it and what new features we are bringing to event-native API Management regarding synchronous and asynchronous APIs.

Reviewer: ```suggestion
WARNING: This new policy execution engine has been released as a BETA version: it should not be used in production. The production-ready GA release, due in 2023, will come with additional features - link:https://www.gravitee.io/contact-us[contact us] if you want to know more about it and what new capabilities and improvements we are bringing to event-native API Management regarding synchronous and asynchronous APIs.
```
